### PR TITLE
Add Feature: Krill Herd

### PR DESCRIPTION
**Pull request by TimothyRogers**
New optimiser class Krill Herd. Implemented from Gandomi, Amir Hossein, and Amir Hossein Alavi. "Krill herd: a new bio-inspired optimization algorithm." Communications in nonlinear science and numerical simulation 17.12 (2012): 4831-4845. 

Added tech: 
- New solution type krill, basically a particle with motion and foraging memory
- Helper function to get bounds as numpy array
- Helper function to linearly decrease parameters over generations, e.g. inertia

Notes:
- The Krill are not a fan of the Ackley function
- Implementation not quite true to the paper, implemented 1-to-gbest crossover not 1-to-rand (fixed Ackley issue)
- Fixed some hyperparameters that are not clear from the paper, e.g. Dmax is not randomly chosen here
- Wasn't clear how to initialise induced motion and foraging, have done so randomly within reasonable values
- There should probably be a bit of thought given to the solution base classes...

Still Needed:
- Update README
- Chaotic/Quantum Krill....

### CODE REVIEW

**Comment by TimothyRogers** [comment]
- Build failed on 5466088 due to lack of scipy. Fixed in 64464c2
- Build failed in 64464c2 from 4D Ackley minimum tolerance. Fixed in 6fdfe7f

File: src/freelunch/optimisers.py [hunk 1]
```diff
@@ -1,145 +1,402 @@
...
+            motion[i,:] = krill.motion
+            forage[i,:] = krill.forage
+
+        return [vals,locs,motion,forage]
+
+
+
+    def winners_and_losers(self,herd):
+        '''
+        Sometimes in life you're the best sometimes you're the worst, sorry krill
```
Comment (MDCHAMP) [inline src/freelunch/optimisers.py:224]: hahaha

File: src/freelunch/optimisers.py [hunk 2]
```diff
@@ -1,145 +1,402 @@
...
+            new_pos = current_herd + dt*V
+
+            # Compute objectives and update the herd
+            for i,(dna,motion,forage) in enumerate(zip(new_pos,N,F)):
+                pop[i].pos = dna
+                pop[i].motion = motion
+                pop[i].forage = forage
+                pop[i].fitness = self.obj(dna)
+
+        return [ p.as_sol() for p in pop ]
```
Comment (MDCHAMP) [inline src/freelunch/optimisers.py:401]: does this return the solutions in fitness order?  I think the API assumes that they are, I'll double check and add a test case for this.

Comment (TimothyRogers) [reply src/freelunch/optimisers.py:401]: Doesn't currently return in order. My bad. Also not in PSO, will open issue.

Comment (MDCHAMP) [reply src/freelunch/optimisers.py]: I think the build will fail until then so will hold off on merging the PR for now

**Comment by MDCHAMP** [comment]
Closes #6  

This looks great :) Can't wait for the penguins to have some lovely krill to eat...

 (I'm not sure the that penguins eat krill but then again dodgy nature analogies are the whole point of this)

I don't have anything on this afternoon so I'll get the SADE implementation in!

**Comment by TimothyRogers** [comment]
Clearly, I have made a hash of the merge... should be good  now

**Pull request merged**